packages/discovery-provider/src/queries/query_helpers.py

```diff
@@ -590,71 +590,103 @@ def _populate_gated_track_metadata(session, tracks, current_user_id):
-    gated_tracks = list(
+    gated_track_access = {
+        getContentId(metadata): defaultdict() for metadata in entities
+    }
+    gated_entities = list(
         filter(
-            lambda track: track.get("stream_conditions")
-            or track.get("download_conditions"),
-            tracks,
+            lambda metadata: metadata.get("stream_conditions")
+            or metadata.get("download_conditions"),
+            entities,
         )
     )
-    gated_track_ids = set([track["track_id"] for track in gated_tracks])
+    gated_content_ids = set([getContentId(metadata) for metadata in gated_entities])
     gated_content_access_args = []
-    for track in gated_tracks:
+    for entity in gated_entities:
         gated_content_access_args.append(
             {
                 "user_id": current_user_id,
-                "content_id": track["track_id"],
-                "content_type": "track",
+                "content_id": getContentId(entity),
+                "content_type": content_type,
             }
         )
 
     gated_content_access = content_access_checker.check_access_for_batch(
         session, gated_content_access_args
     )
 
-    for track in gated_tracks:
-        track_id = track["track_id"]
+    for entity in gated_entities:
+        content_id = getContentId(entity)
+        gated_access = gated_content_access[content_type]
         has_stream_access = (
-            current_user_id in gated_content_access["track"]
-            and track_id in gated_content_access["track"][current_user_id]
-            and gated_content_access["track"][current_user_id][track_id][
-                "has_stream_access"
-            ]
+            current_user_id in gated_access
+            and content_id in gated_access[current_user_id]
+            and gated_access[current_user_id][content_id]["has_stream_access"]
         )
         has_download_access = (
-            current_user_id in gated_content_access["track"]
-            and track_id in gated_content_access["track"][current_user_id]
-            and gated_content_access["track"][current_user_id][track_id][
-                "has_download_access"
-            ]
+            current_user_id in gated_access
+            and content_id in gated_access[current_user_id]
+            and gated_access[current_user_id][content_id]["has_download_access"]
         )
-        gated_track_access[track_id]["has_stream_access"] = has_stream_access
-        gated_track_access[track_id]["has_download_access"] = has_download_access
+        gated_track_access[content_id]["has_stream_access"] = has_stream_access
+        gated_track_access[content_id]["has_download_access"] = has_download_access
 
-    for track in tracks:
-        track_id = track["track_id"]
-        if track_id not in gated_track_ids:
-            track[response_name_constants.access] = {
+    for entity in entities:
+        content_id = getContentId(entity)
+        if content_id not in gated_content_ids:
```

Comment:
i was gonna say i'm scared of id collisions between tracks and playlists, but since we only ever call this endpoint for tracks or playlists separately i think we're ok.